File: internal/net/ws_test.go
package net

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	core "clipsync/internal"

	"nhooyr.io/websocket"
)

// TestWSHandshake verifies WebSocket client sends auth header.
func TestWSHandshake(t *testing.T) {
	var gotAuth string

	// WebSocket server that captures the auth header
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("X-Auth-Token")
		c, err := websocket.Accept(w, r, nil)
		if err != nil {
			t.Fatalf("accept: %v", err)
		}
		defer c.Close(websocket.StatusNormalClosure, "bye")
	}))
	defer ts.Close()

	// convert http:// to ws://
	wsURL := "ws" + ts.URL[4:]

	cli, err := NewWS(wsURL, "deadbeef", "test-secret-key")
	if err != nil {
		t.Fatalf("NewWS: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	out := make(chan core.Snapshot)
	go cli.Poll(ctx, out)

	// wait for connection
	time.Sleep(100 * time.Millisecond)

	if gotAuth == "" {
		t.Fatalf("no auth header received")
	}
}

// TestWSEcho verifies send/receive through WebSocket.
func TestWSEcho(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		c, err := websocket.Accept(w, r, nil)
		if err != nil {
			return
		}
		defer c.Close(websocket.StatusNormalClosure, "")

		// echo server
		for {
			_, msg, err := c.Read(context.Background())
			if err != nil {
				return
			}
			// echo back
			if err := c.Write(context.Background(), websocket.MessageText, msg); err != nil {
				return
			}
		}
	}))
	defer ts.Close()

	wsURL := "ws" + ts.URL[4:]
	cli, _ := NewWS(wsURL, "me", "test-secret-key")

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	out := make(chan core.Snapshot, 1)
	go cli.Poll(ctx, out)

	// wait for connection
	time.Sleep(100 * time.Millisecond)

	// send a snapshot
	want := core.Snapshot{
		Origin: "other",
		Items:  []core.Item{{Fmt: 1, Payload: "dGVzdA=="}},
	}

	if err := cli.Send(want); err != nil {
		t.Fatalf("Send: %v", err)
	}

	// should receive echo (but filtered since Origin != "me")
	select {
	case got := <-out:
		if got.Origin != want.Origin {
			t.Fatalf("origin mismatch")
		}
	case <-time.After(2 * time.Second):
		t.Fatalf("timeout waiting for echo")
	}
}

// TestWSReconnect verifies reconnection behavior.
func TestWSReconnect(t *testing.T) {
	var connCount int

	// server that accepts only first connection
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		connCount++
		if connCount == 1 {
			c, _ := websocket.Accept(w, r, nil)
			// immediately close to trigger reconnect
			c.Close(websocket.StatusNormalClosure, "test")
		} else {
			c, _ := websocket.Accept(w, r, nil)
			// keep second connection alive
			time.Sleep(1 * time.Second)
			c.Close(websocket.StatusNormalClosure, "")
		}
	}))
	defer ts.Close()

	wsURL := "ws" + ts.URL[4:]
	cli, _ := NewWS(wsURL, "me", "test-secret-key")

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	out := make(chan core.Snapshot)
	go cli.Poll(ctx, out)

	// wait for reconnect
	time.Sleep(1500 * time.Millisecond)

	if connCount < 2 {
		t.Fatalf("expected at least 2 connections, got %d", connCount)
	}
}

File: internal/net/ws_test.py
// ws.go — WebSocket transport implementing the Client interface.
// Uses nhooyr.io/websocket.  All writes serialised via a sync.Mutex.
package net

import (
    "context"
    "encoding/json"
    "errors"
    "sync"
    "time"

    core "clipsync/internal"

    "nhooyr.io/websocket"
)

// wsClient keeps one persistent socket; reconnects with back‑off.
type wsClient struct {
    url string
    *shared
    conn *websocket.Conn
    mu   sync.Mutex // serialises all writes (Ping + Send)
}

var _ Client = (*wsClient)(nil)

func NewWS(url, id, keyHex string) (*wsClient, error) {
    sh, err := newShared(id, keyHex)
    if err != nil {
        return nil, err
    }
    return &wsClient{url: url, shared: sh}, nil
}

/*──────────── dial / close helpers ───────────*/
func (c *wsClient) dial(ctx context.Context) error {
    hdr := map[string][]string{"X-Auth-Token": {c.buildAuthHeader()}}
    ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
    defer cancel()
    conn, _, err := websocket.Dial(ctx, c.url, &websocket.DialOptions{HTTPHeader: hdr})
    if err != nil {
        return err
    }
    c.conn = conn
    return nil
}

func (c *wsClient) close() {
    if c.conn != nil {
        _ = c.conn.Close(websocket.StatusNormalClosure, "bye")
        c.conn = nil
    }
}

/*──────────── Client.Send ───────────*/
func (c *wsClient) Send(snap core.Snapshot) error {
    if c.conn == nil {
        return errors.New("ws: not connected")
    }
    snap.Quick = core.QuickKey(snap.Items)
    msg := mustJSON(snap)
    if len(msg) > bodyCap {
        return errors.New("body >32 MiB, dropped")
    }
    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
    defer cancel()

    c.mu.Lock()
    err := c.conn.Write(ctx, websocket.MessageText, msg)
    c.mu.Unlock()
    return err
}

/*──────────── Client.Poll ───────────*/
func (c *wsClient) Poll(ctx context.Context, out chan<- core.Snapshot) {
    backoff := 500 * time.Millisecond
reconnect:
    if err := c.dial(ctx); err != nil {
        select {
        case <-ctx.Done():
            return
        case <-time.After(backoff):
            backoff = minDuration(backoff*2, 8*time.Second)
            goto reconnect
        }
    }
    backoff = 500 * time.Millisecond // reset on success

    ping := time.NewTicker(25 * time.Second)
    defer ping.Stop()

    for {
        select {
        case <-ctx.Done():
            c.close()
            return
        case <-ping.C:
            c.mu.Lock()
            _ = c.conn.Ping(context.Background())
            c.mu.Unlock()
        default:
            _, data, err := c.conn.Read(ctx)
            if err != nil {
                c.close()
                goto reconnect
            }
            if len(data) > bodyCap {
                continue
            }
            var snap core.Snapshot
            if json.Unmarshal(data, &snap) != nil {
                continue
            }
            if snap.Origin != c.id {
                out <- snap
            }
        }
    }
}

func minDuration(a, b time.Duration) time.Duration {
    if a < b {
        return a
    }
    return b
}
